lib/dao.js

```diff
@@ -2407,6 +2586,183 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
 };
 
 /**
+ * Replace set of attributes.
+ * Performs validation before replacing.
+ *
+ * @trigger `validation`, `save` and `update` hooks
+ * @param {Object} data Data to replace
+ * @param {Object} [options] Options for replace
+ * @param {Function} cb Callback function called with (err, instance)
+ */
+DataAccessObject.prototype.replaceAttributes = function(data, options, cb) {
+  var Model = this.constructor;
+  var id = getIdValue(this.constructor, this);
+  return Model.replaceById(id, data, options, cb);
+};
+
+DataAccessObject.replaceById = function(id, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+  assert((options || cb), 'The cb must be passed');
+
+  if (cb === undefined) {
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+  }
+
+  cb = cb || utils.createPromiseCallback();
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+          'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = this.getConnector();
+  assert(typeof connector.replaceById === 'function',
+          'replaceById() must be implemented by the connector');
+
+  var pkName = idName(this);
+  if (!data[pkName]) data[pkName] = id;
+
+  var Model = this;
+  var inst = new Model(data);
+  var enforced = {};
+  this.applyProperties(enforced, inst);
+  inst.setAttributes(enforced);
+  Model = this.lookupModel(data); // data-specific
+  if (Model !== inst.constructor) inst = new Model(data);
+  var strict = inst.__strict;
+
+  if (isPKMissing(Model, cb))
+    return cb.promise;
+
+  var model = Model.modelName;
+  var hookState = {};
+
+  // Convert the data to be plain object so that update won't be confused
+  if (data instanceof Model) {
+    data = data.toObject(false);
+  }
+  data = removeUndefined(data);
+
+  if (id !== data[pkName]) {
+    var err = new Error('id property (' + pkName + ') ' +
+            'cannot be updated from ' + inst[pkName] + ' to ' + data[pkName]);
+    err.statusCode = 400;
+    process.nextTick(function() { cb(err); });
+    return cb.promise;
+  }
+
+  var context = {
+    Model: Model,
+    instance: inst,
+    isNewInstance: false,
+    hookState: hookState,
+    options: options
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
```

Comment:
@bajtos Regarding this [comment]: I can do `copyData(ctx.instance, data);` so that I consider instance changes in `before save`, but I'm ignoring `data` in this case if instance has not been changed in `before save`... Any suggestions?